crates/llms/src/anthropic/chat.rs

```diff
@@ -0,0 +1,397 @@
+                                        code: None,
+                                    })
+                                })?,
+                                call.function.name.clone(),
+                            )))
+                        })
+                        .collect::<Result<_, OpenAIError>>()?,
+                    None => vec![],
+                };
+
+                content_blocks.extend(tool_blocks);
+                Ok(MessageParam::assistant(content_blocks))
+            }
+        }
+    }
+}
+
+impl TryFrom<(AnthropicModelVariant, CreateChatCompletionRequest)> for MessageCreateParams {
+    type Error = OpenAIError;
+    fn try_from(
+        pair: (AnthropicModelVariant, CreateChatCompletionRequest),
+    ) -> Result<Self, Self::Error> {
+        let (model, value) = pair;
+
+        let messages = value
+            .messages
+            .iter()
+            .filter(|m| !matches!(m, ChatCompletionRequestMessage::System(_)))
+            .map(|m| MessageParam::try_from(m.clone()))
+            .collect::<Result<Vec<_>, _>>()?;
+
+        Ok(MessageCreateParams {
+            top_k: value.top_logprobs.map(u32::from),
+            top_p: value.top_p,
+            temperature: value.temperature,
+            max_tokens: value
+                .max_completion_tokens
+                .unwrap_or(model.default_max_tokens()),
+            stream: value.stream,
+            metadata: value
+                .metadata
+                .and_then(|m| m.get("user_id").cloned())
+                .map(|id| MetadataParam {
+                    user_id: id.as_str().map(String::from),
+                }),
+            model,
+            stop_sequences: value.stop.map(|s| match s {
+                Stop::String(s) => vec![s],
+                Stop::StringArray(a) => a,
+            }),
+            system: system_message_from_messages(&value.messages),
+            messages,
+
+            tool_choice: match value.tool_choice {
+                Some(ChatCompletionToolChoiceOption::Auto) => Some(ToolChoiceParam::auto(
+                    !value.parallel_tool_calls.unwrap_or_default(),
+                )),
+                None | Some(ChatCompletionToolChoiceOption::None) => None,
+                Some(ChatCompletionToolChoiceOption::Required) => Some(ToolChoiceParam::any(
+                    !value.parallel_tool_calls.unwrap_or_default(),
+                )),
+                Some(ChatCompletionToolChoiceOption::Named(ChatCompletionNamedToolChoice {
+                    function: FunctionName { name },
+                    ..
+                })) => Some(ToolChoiceParam::tool(
+                    name,
+                    !value.parallel_tool_calls.unwrap_or_default(),
+                )),
+            },
+            tools: value.tools.map(|t| t.iter().map(Into::into).collect()),
+        })
+    }
+}
+
+fn system_message_from_messages(messages: &[ChatCompletionRequestMessage]) -> Option<String> {
```

Comment:
This might be a useful function to have available for other LLM model types?

Reply:
This is primarily needed for LLM providers that use system messages as an alternative field, and not a standalone message. Currently this is just Anthropic.